tools/data-conversion/src/u8vector.rs

```diff
@@ -0,0 +1,181 @@
+// use crate::ir::IntermediateRepresentation;
+use crate::ir::{IntermediateRepresentation, SpecialCase};
+use num_bigint::{BigInt, BigUint};
+use num_traits::One;
+
+pub fn binary_to_u8_vec(binary_string: &str) -> Result<Vec<u8>, String> {
+    // Determine the necessary length to pad to the nearest multiple of 8 bits
+    let total_length = ((binary_string.len() + 7) / 8) * 8;
+
+    // Get the most significant bit; default to '0' if the string is empty
+    let padding_char = binary_string.chars().next().unwrap_or('0');
+
+    // Pad the binary string with the most significant bit
+    let current_len = binary_string.len();
+    let padded_binary_string = if current_len >= total_length {
+        binary_string.to_string()
+    } else {
+        let padding = std::iter::repeat(padding_char)
+            .take(total_length - current_len)
+            .collect::<String>();
+        format!("{}{}", padding, binary_string)
+    };
+
+    // Convert the padded binary string to a u8 vector
+    let mut vec = Vec::new();
+    for chunk in padded_binary_string.as_bytes().chunks(8) {
+        let byte_str = std::str::from_utf8(chunk)
+            .map_err(|_| "Invalid UTF-8 in binary string.")?;
+        let byte = u8::from_str_radix(byte_str, 2)
+            .map_err(|_| "Invalid binary number.")?;
+        vec.push(byte);
+    }
+
+    Ok(vec)
+}
+
+pub fn hex_to_u8_vec(hex: &str) -> Result<Vec<u8>, String> {
+    let mut padded_hex = hex.to_string();
+
+    // If the hex string length is not a multiple of 2, pad with the most significant nibble
+    let padding = 2 - (padded_hex.len() % 2);
+    if padding != 2 {
+        let msb = &padded_hex[0..1]; // Get the most significant nibble
+        padded_hex = msb.repeat(padding) + &padded_hex; // Pad with the MSB
+    }
+
+    let mut vec = Vec::new();
+
+    for i in (0..padded_hex.len()).step_by(2) {
+        let byte_str = &padded_hex[i..i + 2]; // Fixed to extract two hex digits
+        match u8::from_str_radix(byte_str, 16) {
+            Ok(byte) => vec.push(byte),
+            Err(_) => return Err(String::from("Invalid hex string")),
+        }
+    }
+    Ok(vec)
+}
+
+/// Converts a fixed-point number represented as a u8 vector into an IntermediateRepresentation
+///
+///  # Arguments
+///
+/// * `vector` - A vector of u8 values representing the fixed-point number.
+/// * `exponent` - The value of the exponent.
+/// * `twos_comp` - Boolean  indicating whether the fixed-point number uses two's complement for the sign bit.
+///
+/// # Returns
+///
+/// An intermediate representation
+pub fn u8_to_ir_fixed(
+    vector: Result<Vec<u8>, String>,
```

Comment:
I think this function probably shouldn't take a `Result` as an argument. (And same comment with the next function.)